src/tools/rbd_nbd/rbd-nbd.cc

```
@@ -792,55 +803,101 @@ static int parse_imgpath(const std::string &imgpath)
+  }
+  ifs >> pid;
+  return pid;
+}
+
+static int do_list_mapped_devices()
+{
+  bool r;
+  bool should_print = false;
+  int index = 0;
+  int pid = 0;
+  TextTable tbl;
+
+  tbl.define_column("pid", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("pool", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("image", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("snap", TextTable::LEFT, TextTable::LEFT);
+  tbl.define_column("device", TextTable::LEFT, TextTable::LEFT);
+
   while (true) {
-    snprintf(path, sizeof(path), "/dev/nbd%d", m);
-    int nbd = open_device(path);
-    if (nbd < 0)
+    std::string nbd_path = "/sys/block/nbd" + stringify(index);
+    if(access(nbd_path.c_str(), F_OK) != 0) {
       break;
-    if (ioctl(nbd, NBD_SET_SOCK, fd[0]) != 0)
-      cout << path << std::endl;
-    else
-      ioctl(nbd, NBD_CLEAR_SOCK);
-    close(nbd);
-    m++;
-  }
+    }
+    std::string pid_path = nbd_path + "/pid";
+    pid = get_map_pid(pid_path);
 
-  close(fd[0]);
-  close(fd[1]);
+    if(pid > 0) {
+      r = get_mapped_info(pid);
+      if (r < 0) {
+        index++;
+        continue;
+      } 
+      should_print = true;
+      if (snapname.empty()) {
+        snapname = "-";
+      }
+      tbl << pid << poolname << imgname << snapname 
+          << "/dev/nbd" + stringify(index) << TextTable::endrow;
+    }
+
+    index++;
+  }
 
+  if (should_print) {
+    cout << tbl;
+  }
   return 0;
 }
 
-static int rbd_nbd(int argc, const char *argv[])
+static int parse_args(vector<const char*>& args)
 {
-  int r;
-  enum {
-    None,
-    Connect,
-    Disconnect,
-    List
-  } cmd = None;
-
-  vector<const char*> args;
-
-  argv_to_vec(argc, argv, args);
-  md_config_t().parse_argv(args);
-
   std::vector<const char*>::iterator i;
   std::ostringstream err;
 
+  md_config_t().parse_argv(args);
+
   for (i = args.begin(); i != args.end(); ) {
     if (ceph_argparse_flag(args, i, "-h", "--help", (char*)NULL)) {
       usage();
```

Calling `usage()` would be better to do outside. E.g you can return a special error here (e.g. `-ENOENT` or `-ENODATA`) to recognize this case.